### ISSUE TITLE
Explain Unfolding step

### ISSUE THREAD

**Issue by bluesky314** [PeterL1n/BackgroundMattingV2#135]
Had a few questions about what is going on here in the sampling stage in the Refiner:
```
if self.patch_crop_method == 'unfold':
               return x.permute(0, 2, 3, 1) \
                    .unfold(1, size + 2 * padding, size) \
                    .unfold(2, size + 2 * padding, size)[idx[0], idx[1], idx[2]]
```
- https://github.com/PeterL1n/BackgroundMattingV2/blob/4a56223a1cd9b2c2678582513c573debbfc12cae/model/refiner.py#L205

1) x is (bs,c,h,w). Why is it being permuted to (bs,h,w,c) before the unfold? 

2) Generally one unfold across the channel dimension should be able to extract the patches. Why are there two unfolds here? 

3) What is the logic behind size + 2 * padding?

**Comment by PeterL1n** [comment]
We want to make the eventual tensor becomes `[batch, n_patch_y, n_patch_x, channel, patch_size_h, patch_size_w]`. This allow us to use index `[batch, patch_idx_y, patch_idx_x]` to select out tensor patches of shape `[n_patch, channel, patch_size_y, patch_size_x]`.

After the first permute, it becomes `[batch, height, width, channel]`. The first unfold works on height dimension and leave the width dimension unchanged, making the tensor `[batch, n_patch_y, width, channel, patch_size_h]`. The second unfold works on the width dimension, making the tensor `[batch, n_patch_y, n_patch_x, channel, patch_size_h, patch_size_w]`. You need two unfold to work on two spatial dimensions.

The pytorch unfold function takes patch size, and stride size two parameters to allow cropping of overlapping patches. In this method, we define size to be the central non-overlapping area of each crop, and padding as the overlapping extension. so size + 2 * padding becomes the actual width and height of the crop.

**Comment by PeterL1n** [comment]
Some additional thoughts.

The implementation in this repository sticks with the primitive functions PyTorch offers. This means in order to perform convolution only on selective patches, we always need to crop out tensor, convolve, then replace in tensor. This involves additional memory copy overhead.

I think it is possible to write a custom Cuda kernel, given the target patch locations, and the Cuda kernel only convolves the patch region without making any additional tensor copies. It would be interesting if someone wants to experiment it.

**Comment by bluesky314** [comment]
Thanks @PeterL1n for your reply. Will spend some time on this in the coming week and revert if any more doubts.

**Comment by bluesky314** [comment]
Was able to understand how the unfolding works however the idx we have here is the pixel indexes on the original grid. Once we change the grid via unfold to folded regions(with the particular step size) that idx scheme no longer applies. 

eg) 
```
size=3
padding=1
x=torch.rand((1,3,320,320))
out=x.permute(0, 2, 3, 1).unfold(1, size + 2 * padding, size) \
                    .unfold(2, size + 2 * padding, size) 
out.size()
```
`In [9]: torch.Size([1, 106, 106, 3, 5, 5])`

So nonzero points on the original grid of x of 320X320 no longer match to the new grid of 106 X 106.

 According to this the indexing with idx is not correct. Is something being missed here?

**Comment by PeterL1n** [comment]
First of all, you missed a padding op. See https://github.com/PeterL1n/BackgroundMattingV2/blob/4a56223a1cd9b2c2678582513c573debbfc12cae/model/refiner.py#L203

Second, 106 is right because 320 / 3 roughly equals 106. 3 is a bad choice because 320 is not really divisible by 3. Let change it to 2 for example. It should be:
```py
size = 2
padding = 1
x = torch.randn((1, 3, 320, 320))
x = torch.nn.functional.pad(x, (padding, padding, padding, padding))
x = x.permute(0, 2, 3, 1) \
        .unfold(1, size + 2 * padding, size) \
        .unfold(2, size + 2 * padding, size)
print(x.shape) # torch.Size([1, 160, 160, 3, 4, 4])
```
To select patches, you will need a index of 160x160:

```py
idx = torch.rand((1, 160, 160)).gt(0.5) # bool tensor
x = x[idx] # shape: (num_patches, 3, 4, 4)
```
The above code calculates nonzeros internally at `x[idx]`, if you need to use `idx` to select multiple times, it is faster to calculate nonzeros yourself ones and reuse it elsewhere.

```py
idx = torch.rand((1, 160, 160)).gt(0.5) # bool tensor
idx = torch.nonzero(idx) # shape: (num_patches, 3)
x = x[idx[:, 0], idx[:, 1], idx[:, 2]] # shape: (num_patches, 3, 4, 4)
```

**Comment by bluesky314** [comment]
Thanks @PeterL1n, yes totally missed the importance of the padding step. Will go through this more throughly and reply with further questions.
